src/lib/support/logging/Trace.h

```diff
@@ -0,0 +1,78 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/support/logging/CHIPLogging.h>
+
+namespace chip {
+namespace Logging {
+
+/// Provides the ability to add tracing code to methods. Will do
+/// ChipLog tracing if CHIP_CONFIG_ENABLE_FUNCT_ERROR_LOGGING is set
+/// and will be optimized away by the compiler if the define is not set
+///
+/// Typical usage:
+///     LogTracer tracer(__func__);
+///
+///     VerifyOrReturnError(somecondition, tracer.log(CHIP_ERROR_INVALID_ARGUMENT));
+///     VerifyOrReturnError(otherCondition, tracer.log(CHIP_ERROR_INVALID_ARGUMENT, "othercondition"));
+///
+///     ReturnTracedErrorOnFailure(foo.Bar(), tracer);
+///     ReturnTracedErrorOnFailure(foo.Baz(), tracer);
+///
+///     return CHIP_NO_ERROR;
+class LogTracer
+{
+#if CHIP_CONFIG_ENABLE_FUNCT_ERROR_LOGGING
+public:
+    LogTracer(const char * name) : mName(name) {}
+    inline CHIP_ERROR logError(CHIP_ERROR err)
```

Comment:
You don't need to say `inline` here.

Reply:
Also, this is the wrong naming style.

Reply:
> I would suggest just removing these; ChipError can record the exact line that caused it now, which probably obviates the need for them.
> 
> If we're concerned about callers not checking the error there is [[nodiscard]] and non-standard equivalents.

I would very much prefer this, however I was afaraid of not losing anything that ChipLogFunctError was providing.

Overall I find ChipLogFunctError somewhat buggy: it  gives me the __LINE__ for where the function is called and NOT the location of chiperror. Many methods return INVALID_STATE or ILLEGAL_ARGUMENT over many paths and  knowing one of the arguments (out of many in some cases) is bad does not help as much.

What the function does give in the current state is somewhat of a stack trace (like knowing 'clusterid was invalid' is interesting, but knowing who called things with an invalid cluster could be useful.

Thoughts? Could I get some +1 or "please don't" on the idea of  'just remove ChipLogFunctError alltogether'?

Reply:
Updated the inline/style here, resolving conversation to make things look "greener" but will extract the "Why don't we just remove things" conversation separately.